packages/world/ts/node/render-solidity/renderSystemLibrary.ts

```diff
@@ -0,0 +1,260 @@
+      ${renderReturnParameters(returnParameters)}
+  `;
+
+  const rootSystemCheck = `
+    // if the contract calling this function is a root system, it should use \`callAsRoot\`
+    if (address(_world()) == address(this)) revert ${callingFromRootSystemErrorName}();
+  `;
+
+  const encodedSystemCall = renderEncodeSystemCall(systemLabel, name, parameters);
+
+  if (stateMutability === "") {
+    return `
+      ${functionSignature} {
+        ${rootSystemCheck}
+        bytes memory systemCall = ${encodedSystemCall};
+        bytes memory result = self.from == address(0) ? _world().call(self.systemId, systemCall) : _world().callFrom(self.from, self.systemId, systemCall);
+        ${renderAbiDecode(returnParameters)}
+      }
+    `;
+  } else {
+    return `
+      ${functionSignature} {
+        ${rootSystemCheck}
+        bytes memory systemCall = ${encodedSystemCall};
+        bytes memory worldCall = self.from == address(0)
+          ? abi.encodeCall(IWorldCall.call, (self.systemId, systemCall))
+          : abi.encodeCall(IWorldCall.callFrom, (self.from, self.systemId, systemCall));
+        (bool success, bytes memory returnData) = address(_world()).staticcall(worldCall);
+        if (!success) revertWithBytes(returnData);
+        bytes memory result = abi.decode(returnData, (bytes));
+        ${renderAbiDecode(returnParameters)}
+      }
+    `;
+  }
+}
+
+function renderRootCallWrapperFunction(contractFunction: ContractInterfaceFunction, systemLabel: string) {
+  const { name, parameters, stateMutability, returnParameters } = contractFunction;
+
+  const functionArguments = ["RootCallWrapper memory self", ...parameters];
+
+  const functionSignature = `
+    function ${name}(
+      ${renderArguments(functionArguments)}
+    ) internal
+      ${stateMutability === "pure" ? "view" : stateMutability}
+      ${renderReturnParameters(returnParameters)}
+  `;
+
+  const encodedSystemCall = renderEncodeSystemCall(systemLabel, name, parameters);
+
+  if (stateMutability === "") {
+    return `
+      ${functionSignature} {
+        bytes memory systemCall = ${encodedSystemCall};
+        bytes memory result = SystemCall.callWithHooksOrRevert(self.from, self.systemId, systemCall, msg.value);
+        ${renderAbiDecode(returnParameters)}
+      }
+    `;
+  } else {
+    return `
+      ${functionSignature} {
+        bytes memory systemCall = ${encodedSystemCall};
+        bytes memory result = SystemCall.staticcallOrRevert(self.from, self.systemId, systemCall);
+        ${renderAbiDecode(returnParameters)}
+      }
+    `;
+  }
+}
+
+function renderEncodeSystemCall(interfaceName: string, functionName: string, parameters: string[]) {
+  const paramNames = parameters.map((param) => param.split(" ").slice(-1)[0]).join(", ");
+  return `abi.encodeCall(${interfaceName}.${functionName}, (${paramNames}))`;
+}
+
+function renderAbiDecode(returnParameters: string[]) {
+  if (returnParameters.length === 0) return "result;";
```

Comment:
should this be `return result;`? 🤔 

```suggestion
  if (returnParameters.length === 0) return "return result;";
```

Reply:
It is so that the compiler doesn't complain about an unused variable. I guess we could just not assign it if not needed

Reply:
fixed by only assigning the result if needed